_posts/2023-11-30-prevent-broker-scale-down-if-containing-paritition-replicas.md

```diff
@@ -0,0 +1,182 @@
+spec:
+  kafka:
+    version: 3.6.0
+    replicas: 3 # Changes the replicas to 3
+    listeners:
+# ....    
+```
+
+Now you can apply the updated Kafka custom resource and check if the brokers are scaled down or not.
+
+You check the status of the Kafka custom resource using this command:
+```sh
+kubectl get kafka my-cluster -n myproject -o yaml
+```
+
+Since we didn't move the replicas from the broker to be removed, the scale down fails and, as indicated by the errors given in the status of the Kafka custom resource:
+```yaml
+status:
+  clusterId: S4kUmhqvTQegHCGXPrHe_A
+  conditions:
+  - lastTransitionTime: "2023-12-04T07:03:49.161878138Z"
+    message: Cannot scale down brokers [3] because brokers [3] are not empty
+    reason: InvalidResourceException
+    status: "True"
+    type: NotReady
+```
+So these logs are basically telling you that broker 3 is not empty, which makes the whole reconciliation fail. You can get rid of this error by reverting `.spec.kafka.replicas` in the Kafka resource.
+
+### Scaling down after moving out all replicas from the broker to be removed
+
+Let's try to scale down the broker now after emptying partition replicas from the broker to be removed.
+
+We can make use of Cruise Control's `KafkaRebalance` resource in Strimzi with `remove-broker` node configuration for this job.
+Here is an example `KafkaRebalance` resource:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalance
+metadata:
+  name: my-rebalance
+  labels:
+    strimzi.io/cluster: my-cluster
+# no goals specified, using the default goals from the Cruise Control configuration
+spec:
+  mode: remove-brokers
+  brokers: [3]
+```
+You can create this `KafkaRebalance` custom resource and this will allow Cruise Control to handle the task of rebalancing and moving the partition replicas from the broker that is going to be removed.
+For more in-depth information you can refer to our [Rebalancing cluster using Cruise Control](https://strimzi.io/docs/operators/latest/deploying#proc-generating-optimization-proposals-str) documentation.
+
+Once the rebalacing is done, you can check if the topics are moved from the broker or not:
+```shell
+Topic: my-topic	TopicId: bbX7RyTSSXmheaxSPyRIVw	PartitionCount: 3	ReplicationFactor: 3	Configs: min.insync.replicas=2,segment.bytes=1073741824,retention.ms=7200000,message.format.version=3.0-IV1
+	Topic: my-topic	Partition: 0	Leader: 2	Replicas: 2,0,1	Isr: 2,1,0
+	Topic: my-topic	Partition: 1	Leader: 2	Replicas: 2,1,0	Isr: 1,0,2
+	Topic: my-topic	Partition: 2	Leader: 1	Replicas: 1,0,2	Isr: 1,0,2
+```
+
+Now you can scale down the broker, and it will happen seamlessly since the broker is empty.
+
+## What's next
+
+We hope this blog post has provided you with a clear understanding of how the broker scale-down check operates.
+In upcoming Strimzi releases, we aim to enhance this process by automatically reverting `.spec.kafka.replicas` to its original value if the broker scale-down check fails.
```

Comment:
Is there general agreement on that? I don't remember any place where we update the spec, but maybe I'm wrong.

Reply:
This is a good point from Federico ...
* This will likely not be in the next Strimzi release
* We do not really plan to revert the `.spec.kafka.replicas` => even if we did it is not used with KRaft. I would word it as something like _continue the reconciliation without doing the scaledown_ or something in that sense which is what we plan.